app/demo-geant-integration/HepMC3Reader.cc

```diff
@@ -0,0 +1,162 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2022 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file demo-geant-integration/HepMC3Reader.cc
+//---------------------------------------------------------------------------//
+#include "HepMC3Reader.hh"
+
+#include <mutex>
+#include <G4SystemOfUnits.hh>
+#include <HepMC3/ReaderFactory.h>
+
+#include "corecel/Assert.hh"
+#include "corecel/io/Logger.hh"
+
+#include "GlobalSetup.hh"
+
+namespace
+{
+// Mutex visible to all threads
+static std::mutex mutex;
+} // namespace
+
+namespace demo_geant
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Singleton declaration.
+ */
+static HepMC3Reader* hepmc3_reader_singleton = nullptr;
+
+//---------------------------------------------------------------------------//
+/*!
+ * Return non-owning pointer to a singleton.
+ */
+HepMC3Reader* HepMC3Reader::instance()
+{
+    {
+        mutex.lock();
+        if (!hepmc3_reader_singleton)
+        {
+            hepmc3_reader_singleton = new HepMC3Reader();
+        }
+        mutex.unlock();
+    }
+
+    CELER_LOG_LOCAL(status) << "Instancing HepMC3 singleton";
+    return hepmc3_reader_singleton;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Add HepMC3 primaries to a Geant4 event. This function is called by
+ * `G4VUserPrimaryGeneratorAction::GeneratePrimaries`.
+ */
+void HepMC3Reader::GeneratePrimaryVertex(G4Event* g4_event)
+{
+    mutex.lock();
+
+    // Populate event_primaries
+    auto result = this->store_primaries();
+    CELER_ASSERT(result);
+
+    // Add primaries to event
+    for (const auto& primary : event_primaries_)
+    {
+        // TODO: Do we need to check if vertex is inside world volume?
+
+        // All primaries start at t = 0
+        auto g4_vtx = std::make_unique<G4PrimaryVertex>(
+            primary.vertex[0], primary.vertex[1], primary.vertex[2], 0);
+
+        g4_vtx->SetPrimary(new G4PrimaryParticle(primary.pdg,
+                                                 primary.momentum[0],
+                                                 primary.momentum[1],
+                                                 primary.momentum[2],
+                                                 primary.energy));
+
+        g4_event->AddPrimaryVertex(g4_vtx.release());
+    }
+    const int num_vertices = g4_event->GetNumberOfPrimaryVertex();
+    CELER_ENSURE(num_vertices);
```

Comment:
Since semantically we're testing for a number quantity and not a boolean value...
```suggestion
    CELER_ENSURE(g4_event->GetNumberOfPrimaryVertex() > 0);
```